### DISCUSSION TITLE
syntax for axioms

### DISCUSSION THREAD

**Discussion by ko56** [shop-planner/shop3#96] [Ideas]
I propose a modification to the syntax of axioms, with the same goal as the proposed modifications to the syntax of methods: to make it clearer and less susceptible to errors.

I spent hours today debugging an axiom of the form

```
(:- (complicated-axiom ?x ?y)
   (and (long-conjunction))
)
```
that was returning `?y` as `?<tmp-name>` for no reason that I could see,  only to find out that in the course of editing it, adding and removing `println` statements, etc, the final conjuct `(assign ?y ?<expr>)` had mistakenly ended up outside the `(and (...))`.
So something like a `:condition`, `:cond`, `:case`, `:branch`, ... in front of each branch of the axiom would be very helpful.

P.S. I now see that the problem also involves implicit conjuctions.

**Comment by rpgoldman** [comment]
This could be an alternative syntax, but I wouldn't want to make it the only one. The current syntax is a lispification of Prolog Horn clauses, and so is somewhat natural (to those who know Prolog).  But it wouldn't be hard to add optional markers like the ones you suggest, and if it would be helpful, we can do that.

**Reply by ko56** [reply]
OK, my favorite marker is now `:if`. I think it expresses the semantics well.
I have nothing against Prolog :-),  but the current syntax looks like a long implicit conjunction
(unless you use the optional name_i). 
In any case, if implicit conjunctions were disallowed in SHOP3, the problem I mentioned above would likely not have occurred.
